### PR TITLE
[IGA] 5 parametric shell with unit directors

### PR DESCRIPTION
**Pull request by rath3t**
**Description**
I'm adding here a new shell formulation. Work in progress.

The shell corresponds to a non-linear Reissner-Mindlin formulation with non-linear transversal shear.

### CODE REVIEW

**Comment by rath3t** [comment]
- [ ] 2D Variable Core Regestration
- [ ] subrange and array_1d usage
- [ ] Reference Director construction process
- [ ] Tests

**Comment by rath3t** [comment]
## Description of Director init process.

For the reference geometry we need to solve one sparse system of equation of the size:

nxn * nx 1 = nx3, A*x=b

where n are the number of nodes/controlpoints. The solution of this system of equation will give us the director at each nod in such a way that the interpolated directors at the integrations points are as normal as possible.

We need:
- [ ] function which calculates the elementwise/integration point wise contribution to A and b
- [ ] Nodal Assembly into large Matrix 
- [ ] Save solution at nodes

File: applications/IgaApplication/custom_utilities/director_utilities.cpp [hunk 1]
```diff
@@ -36,17 +36,70 @@ namespace Kratos
...
+        {
+            const auto& ele = mrModelPart.pGetElement(i);
+            const auto& eleGeometry = ele->GetGeometry();
+
+            const SizeType numNodes = eleGeometry.size();
+            RhsEle = zero_matrix<double>(numNodes,3);
+            NTNele = zero_matrix<double>(numNodes);
+
+            Nele = eleGeometry.ShapeFunctionsValues();
+            const SizeType r_number_of_integration_points = ele->GetGeometry().IntegrationPointsNumber();
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_utilities/director_utilities.cpp]:
```suggestion
            const SizeType r_number_of_integration_points = eleGeometry.IntegrationPointsNumber();
```

File: applications/IgaApplication/custom_utilities/director_utilities.cpp [hunk 2]
```diff
@@ -36,17 +36,70 @@ namespace Kratos
...
+            for (SizeType iP = 0; iP < r_number_of_integration_points; ++iP)
+            {
+                const Vector& Nip = row(Nele, iP);
+                const array_1d<double, 3> A3 = ele->GetGeometry().UnitNormal(iP);
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_utilities/director_utilities.cpp]:
```suggestion
                const array_1d<double, 3> A3 = eleGeometry.UnitNormal(iP);
```

File: applications/IgaApplication/custom_utilities/director_utilities.cpp (hunk 1 above)
Comment (rath3t) [reply applications/IgaApplication/custom_utilities/director_utilities.cpp]: can suggestions nicht commiten das ist aber ok , oder?

File: applications/IgaApplication/custom_utilities/director_utilities.cpp (hunk 2 above)
Comment (rath3t) [reply applications/IgaApplication/custom_utilities/director_utilities.cpp]: can suggestions nicht commiten das ist aber ok , oder?

**Comment by rath3t** [comment]
Using the work around for the director update the shell is now working.

The last missing ingredients are:
### High priority
- [ ] Write tests @tteschemacher can you construct a framework for this for a test where the stiffness matrix is testet and a fully non-linear solution test?
- [ ] Test if the moment load processes are working correctly
- [ ] Clean up code

### Low priority
- [ ] The reference director construction can be improved, see comment above

File: applications/IgaApplication/custom_elements/shell_5p_element.cpp [hunk 3]
```diff
@@ -22,7 +22,7 @@ namespace Kratos
 	///@name Initialize Functions
 	///@{
 
-	void Shell5pElement::Initialize(const ProcessInfo& rCurrentProcessInfo)
+	void Shell5pElement::Initialize()
```
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_5p_element.cpp]: this change is wrong, now it will not compile on master

Comment (rath3t) [reply applications/IgaApplication/custom_elements/shell_5p_element.cpp]: Ah ok thank you. I changed it to get the same test call in 

https://github.com/KratosMultiphysics/Kratos/blob/678a0e22ca7b80cb3dea4650ede7c9c8d9ade449/applications/IgaApplication/tests/cpp_tests/test_iga_shell_5p.cpp#L65

as in 

https://github.com/KratosMultiphysics/Kratos/blob/678a0e22ca7b80cb3dea4650ede7c9c8d9ade449/applications/IgaApplication/tests/cpp_tests/test_iga_shell_3p.cpp#L64

but i didn't saw that it change in master

https://github.com/KratosMultiphysics/Kratos/blob/80cf9d9d796d5072b5bdc3c9aa0c012fa07c04c2/applications/IgaApplication/tests/cpp_tests/test_iga_shell_3p.cpp#L66

Thanks!

**Review by philbucher — COMMENTED** [review]
Why do you add final?
Usually we don’t do that

File: applications/IgaApplication/custom_utilities/director_utilities.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            auto geometry = mrModelPart.GetRootModelPart().GetGeometry((IndexType)brep_ids[i]);
+
+            size_t number_of_control_points = geometry.size();
+
+            std::vector<IndexType> eqID(number_of_control_points);
+            for (SizeType inodes = 0; inodes < number_of_control_points; ++inodes)
+                eqID[inodes] = geometry[inodes].GetId();
+
+            SparseMatrixType NTN(number_of_control_points, number_of_control_points, number_of_control_points*4); //inital guess how much non-zero are there
+            // Can't we solve each patch independently? -> much more efficient
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_utilities/director_utilities.cpp]: If we can, we should definitely do it. having multiple solvers is no issue at all.

**Comment by rath3t** [comment]
> Why do you add final?
> Usually we don’t do that

I usually want to perhibit further inheritance since inheritance of a leaf class is usually considered bad, see https://youtu.be/32tDTD9UJCE?t=1377 and the minutes before. Therefore, i also add it to the functions. Furthermore, marking it final the compiler can maybe resolve the vtable calls at compile time.

On the other hand if i dig into the Core Guide Guidelines, i found 

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c128-virtual-functions-should-specify-exactly-one-of-virtual-override-or-final

> Note: On a class defined as final, it doesn’t matter whether you put override or final on an individual virtual function.
Therefore, if the class is final it doesn't matter.

Conclusion: I want to stick to the final of the class to preclude bad inheritance patterns.
For the member functions it doesn't matter, therefore i can change it back to fit the usual code style.

What do you think about it?

File: applications/IgaApplication/custom_elements/shell_5p_element.cpp [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+        const double txdtd1 = inner_prod(t, dtd1);
+        const double txdtd2 = inner_prod(t, dtd2);
+        const double a1xdtd1 = inner_prod(a1, dtd1);
+        const double a2xdtd2 = inner_prod(a2, dtd2);
+        const double a1xdtd2 = inner_prod(a1, dtd2);
+        const double a2xdtd1 = inner_prod(a2, dtd1);
+        const Matrix3d a1dyadt = outer_prod(a1, t);
+        const Matrix3d a2dyadt = outer_prod(a2, t);
+        const Matrix3d dtd1dyadt = outer_prod(dtd1, t);
+        const Matrix3d dtd2dyadt = outer_prod(dtd2, t);
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_elements/shell_5p_element.cpp]: those are the terms which make problems within the complete line

**Review by tteschemacher — DISMISSED** [review]
Hey @dagmawibekel and @manuelmessmer ,

if you want to have a check at the shell formulation coming from @rath3t . It is based upon a director on each control point. The shell is pretty fast and can run with a polynomial degree of 1 which is pretty nice for CAD stuff. If you want you can have a look at it before we merge. However, it will be a bit work in progress as well, so if there are things we can change that later as well.

In either way many thanks @rath3t for the implementation.

**Comment by tteschemacher** [comment]
maybe also @oberbichler, if you are interested.

**Review by tteschemacher — APPROVED** [review]
sweet. Thank you @rath3t for your nice contribution!

**Comment by tteschemacher** [comment]
🥳

**Comment by rath3t** [comment]
Thanksy also here for your help! 

🥳

**Pull request merged**